### PR TITLE
Support Asset payment

### PR DESCRIPTION
**Pull request by echevrier**
- According to subxt, add ExtrinsicParams and BaseExtrinsicParamsBuilder to configure the signed extra. The tip type can be defined for the API.
- If no extra params are defined, the default one will be used. Default values are defined as in subxt
- Let the GenericExtras as it was.
- Some examples don't run locally: get_blocks, benchmark and compose xt offline. These don't run in master too

### CODE REVIEW

File: src/std/mod.rs [hunk 1]
```diff
@@ -508,7 +518,7 @@ where
         xthex_prefixed: String,
         exit_on: XtStatus,
     ) -> ApiResult<Option<Hash>> {
-        debug!("sending extrinsic: {:?}", xthex_prefixed);
+        //println!("sending extrinsic: {:?}", xthex_prefixed);
```
Comment (clangenb) [inline src/std/mod.rs:521]: Please change that back before merge

**Review by clangenb — COMMENTED** [review]
Some minor comments, but it looks good all in all!

File: src/std/rpc/ws_client/mod.rs [hunk 2]
```diff
@@ -61,18 +62,22 @@ pub trait Subscriber {
...
     }
 }
 
-impl<P, Client> Api<P, Client>
+impl<P, Client, Params> Api<P, Client, Params>
 where
     P: Pair,
     MultiSignature: From<P::Signature>,
     Client: RpcClientTrait + Subscriber,
+    Params: ExtrinsicParams,
```
[the remark this answers is not in the document]
Comment (clangenb) [reply src/std/rpc/ws_client/mod.rs:80]: Ah I see, makes sense! Then it is ok!

File: compose-macros/src/lib.rs [hunk 3]
```diff
@@ -62,17 +62,22 @@ macro_rules! compose_extrinsic_offline {
...
+            SignedPayload, UncheckedExtrinsicV4,
         };
         use $crate::sp_runtime::{generic::Era, traits::IdentifyAccount, MultiSigner};
 
-        let extra = GenericExtra::new($era, $nonce);
+        let other_params = $extrinsic_params.unwrap_or_default();
+
+        let params = BaseExtrinsicParams::new($nonce, other_params);
+
+        let extra = GenericExtra::from(params);
```
[the remark this answers is not in the document]
Comment (echevrier) [reply compose-macros/src/lib.rs:80]: OK

File: src/extrinsic/balances.rs [hunk 4]
```diff
@@ -36,18 +39,23 @@ pub type BalanceSetBalanceFn = (
     Compact<Balance>,
 );
 
+pub type BalancesExtrinsicParamsBuilder = PlainTipExtrinsicParamsBuilder;
```
[the remark this answers is not in the document]
Comment (echevrier) [reply src/extrinsic/balances.rs]: Too bad. I will remove it

File: src/extrinsic/balances.rs [hunk 5]
```diff
@@ -36,18 +39,23 @@ pub type BalanceSetBalanceFn = (
...
     P: Pair,
     MultiSignature: From<P::Signature>,
     MultiSigner: From<P::Public>,
     Client: RpcClient,
+    Params: ExtrinsicParams,
 {
-    pub fn balance_transfer(&self, to: GenericAddress, amount: Balance) -> BalanceTransferXt {
+    pub fn balance_transfer(&self, to: GenericAddress, amount: Balance) -> BalanceTransferXt
+    where
+        Params: ExtrinsicParams<OtherParams = BalancesExtrinsicParamsBuilder>,
```
[the remark this answers is not in the document]
Comment (echevrier) [reply src/extrinsic/balances.rs]: Because of the call of compose_extrinsic! we have to specify the OtherParams. But yes it should be generic.

File: src/extrinsic/contract.rs [hunk 6]
```diff
@@ -48,20 +51,26 @@ pub type ContractInstantiateFn = (CallIndex, Endowment, GasLimit, Hash, Data);
...
     P: Pair,
     MultiSignature: From<P::Signature>,
     MultiSigner: From<P::Public>,
     Client: RpcClient,
+    Params: ExtrinsicParams,
 {
-    pub fn contract_put_code(&self, gas_limit: Gas, code: Data) -> ContractPutCodeXt {
+    pub fn contract_put_code(&self, gas_limit: Gas, code: Data) -> ContractPutCodeXt
+    where
+        Params: ExtrinsicParams<OtherParams = ContractExtrinsicParamsBuilder>,
```
[the remark this answers is not in the document]
Comment (echevrier) [reply src/extrinsic/contract.rs]: As above

File: src/extrinsic/contract.rs [hunk 7]
```diff
@@ -54,12 +57,16 @@ pub type ContractInstantiateWithCodeXt = UncheckedExtrinsicV4<ContractInstantiat
...
 #[cfg(feature = "std")]
-impl<P, Client> Api<P, Client>
+impl<P, Client, Params, Tip> Api<P, Client, Params>
 where
     P: Pair,
     MultiSignature: From<P::Signature>,
     MultiSigner: From<P::Public>,
     Client: RpcClient,
+    Params: ExtrinsicParams,
+    Params: ExtrinsicParams<OtherParams = BaseExtrinsicParamsBuilder<Tip>>,
```
Comment (clangenb) [inline src/extrinsic/contract.rs:67]: I am confused, why is this line here? 

`Params: ExtrinsicParams` should be enough, no?

**Review by clangenb — COMMENTED** [review]
One minor question, looks good otherwise

File: src/extrinsic/contract.rs (hunk 7 above)
Comment (clangenb) [reply src/extrinsic/contract.rs:67]: Ah ok, I see this is needed, but you can remove the `Params: ExtrinsicParams` because `Params: ExtrinsicParams<OtherParams = BaseExtrinsicParamsBuilder<Tip>` implies the `Params: ExtrinsicParams`.

**Review by clangenb — APPROVED** [review]
Perfect!
